src/image.ts

```diff
@@ -72,19 +73,24 @@ const sign = (salt: string, path: string): string =>
     createHash('md5').update(salt + path).digest('hex');
 
 function src(salt: string, input: string, width: number, dpr: Dpr): string {
-    const url = new URL(input);
-    const service = getSubdomain(url.hostname);
-
-    const params = new URLSearchParams({
-        width: width.toString(),
-        quality: dpr === Dpr.Two ? lowerQuality.toString() : defaultQuality.toString(),
-        fit: 'bounds',
-    });
-
-    const path = `${url.pathname}?${params.toString()}`;
-    const sig = sign(salt, path);
+    const url: Result<string, URL> = fromUnsafe(() => new URL(input), 'invalid url')
+    return pipe2(
+        toOption(url),
+        map(url => {
+            const service = getSubdomain(url.hostname);
+
+            const params = new URLSearchParams({
+                width: width.toString(),
+                quality: dpr === Dpr.Two ? lowerQuality.toString() : defaultQuality.toString(),
+                fit: 'bounds',
+            });
 
-    return `${imageResizer}/${service}${path}&s=${sig}`;
+            const path = `${url.pathname}?${params.toString()}`;
+            const sig = sign(salt, path);
+            return `${imageResizer}/${service}${path}&s=${sig}`;
+    }),
+        withDefault(input)
+    )
```

Comment:
Perhaps it doesn't make too much difference in this case, but worth noting that you don't need to go via `Option` to achieve this. `Result` also has `map` and a way of extracting the value via `either`.

Using `pipe2`:

```ts
return pipe2(
  maybeUrl,
  resultMap(url => {
    const service = getSubdomain(url.hostname);

    const params = new URLSearchParams({
      width: width.toString(),
      quality: dpr === Dpr.Two ? lowerQuality.toString() : defaultQuality.toString(),
      fit: 'bounds',
    });

    const path = `${url.pathname}?${params.toString()}`;
    const sig = sign(salt, path);

    return `${imageResizer}/${service}${path}&s=${sig}`;
  }),
  either(_ => input, url => url),
);
```

Alternatively, using `switch` (we should expose `ResultKind` as we did for `OptionKind`):

```ts
switch (maybeUrl.kind) {
  case ResultKind.Ok:
    const url = maybeUrl.value;
    const service = getSubdomain(url.hostname);

    const params = new URLSearchParams({
      width: width.toString(),
      quality: dpr === Dpr.Two ? lowerQuality.toString() : defaultQuality.toString(),
      fit: 'bounds',
    });

    const path = `${url.pathname}?${params.toString()}`;
    const sig = sign(salt, path);

    return `${imageResizer}/${service}${path}&s=${sig}`;

  case ResultKind.Err:
  default:
    return input;
);
```

Reply:
I think both suggestions look sound to me, as it was a pairing task with @dskamiotis I might ask him what makes more sense for his understanding as to what direction we take.